pkg/recorder/recorder.go

```diff
@@ -21,11 +21,15 @@ limitations under the License.
 package recorder
 
 import (
+	"k8s.io/client-go/tools/events"
 	"k8s.io/client-go/tools/record"
 )
 
 // Provider knows how to generate new event recorders with given name.
 type Provider interface {
 	// NewRecorder returns an EventRecorder with given name.
-	GetEventRecorderFor(name string) record.EventRecorder
+	GetEventRecorder(name string) events.EventRecorder
+	// GetOldEventRecorder returns an EventRecorder for the old events API.
+	// The old API is not 100% supported anymore, use the new one whenever possible.
+	GetOldEventRecorder(name string) record.EventRecorder
```

Comment:
Rather than extending the interface here, would it be possible to build a wrapper that makes a `record.EventRecorder` out of a `events.EventRecorder`?

Reply:
I think there are two main ways we could go about this:
1. We make this a non-breaking change by keeping the existing `GetEventRecorderFor` with the existing signature and add a new `GetEventRecorder` with the new `events.EventRecorder` as return but mark the `GetEventRecorderFor` as `// deprecated`. At some point in a future release we will then remove `GetEventRecorderFor`
2. We make this a breaking change by changing the signature of the existing `GetEventRecorderFor` to return an `events.EventRecorder` and we provide an adapter somewhere to go from `events.EventRecorder` to a `record.EventRecorder` for places like the leader election

My vote if this is possible would be to do 1) as this is the least disruptive for c-r users but its possible I've missed something and

Reply:
Option 1 is doable, but I have to see if the wrapper approach is good enough. Reason being, that the wrapper calls the new API under the hood and both APIs are quite different.  The old API has `AnnotatedEventf` while the new one does not and the new one has things like `relatedObject` and `action`.

Otherwise, the other way to support both is having 2 distinct implementations in separate packages and duplicate the injection and so on on the rest of the codebase.

I will test if the wrapper is viable, present the results and then we can make an informed decision.

Reply:
I have tested out the wrapper and it has the one downside: we loose the ability to set annotations on events since there is no equivalent `AnnotatedEventf`. I also have tried to get some info on why that is the case.

Seeing that, the wrapper can not provide the same features as the old API has, so the only option to support both is probably having both implementations at the same time.

@alvaroaleman my proposal is to duplicate the internal events package and have both implementations separate from each other even if that means lots of duplication for some time. It will make it easier later on to cleanup the old API when removed. What do you think?

Reply:
> @alvaroaleman my proposal is to duplicate the internal events package and have both implementations separate from each other even if that means lots of duplication for some time. It will make it easier later on to cleanup the old API when removed. What do you think?

Not great, but given its for a limited time only that seems okay to me

Reply:
@alvaroaleman I have gone through this and now this PR supports both the old and new events APIs with as little duplication as possible.

There is still a fair amount of duplication but it is confined mostly to internal code and should still be relatively easy to cleanup later on.

If you agree with this approach I will move to adding tests for the new API and move this out of draft.

Reply:
Tests added and moved out of draft. PTAL!